<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [16]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import psycopg2


In [17]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [18]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [19]:
# текст запроса
query = """
    SELECT COUNT(*) AS vacancy_count
    FROM public.vacancies;
"""

In [20]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(f"Количество вакансий в базе: {df['vacancy_count'][0]}")

# Закрываем соединение
connection.close()

Количество вакансий в базе: 49197


C:\Users\zakha\AppData\Local\Temp\ipykernel_14364\3765153580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [21]:
# текст запроса
query = """
    SELECT COUNT(DISTINCT id) AS employer_count
    FROM public.employers;
"""


In [22]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(f"Количество работодателей в базе: {df['employer_count'][0]}")

# Закрываем соединение
connection.close() 

C:\Users\zakha\AppData\Local\Temp\ipykernel_14364\323419175.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


InterfaceError: connection already closed

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
# текст запроса
query = """
    SELECT COUNT(DISTINCT id) AS area_count
    FROM public.areas;
"""


In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(f"Количество регионов в базе: {df['area_count'][0]}")

# Закрываем соединение
connection.close() 

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
# текст запроса
query = """
    SELECT COUNT(DISTINCT id) AS industry_count
    FROM public.industries;
"""


In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(f"Количество сфер деятельности в базе: {df['industry_count'][0]}")

# Закрываем соединение
connection.close()

NameError: name 'connection' is not defined

***

# **Выводы по предварительному анализу данных:**
**Размер базы данных:**

1. База данных достаточно большая и содержит 49197 вакансий, 23501 работодателя, 1362 регионов и 294 сфер деятельности.

**Разнообразие данных:**

1. База данных демонстрирует значительное разнообразие как по количеству вакансий и работодателей, так и по географическому охвату (1362 региона) и сферам деятельности (294). Это позволяет проводить глубокий анализ и получать ценные инсайты.
Потенциальные возможности анализа:

**Анализ вакансий:**
1. Определение наиболее популярных сфер деятельности, регионов, зарплат.
2. Изучение тенденций на рынке труда (сезонность, спрос, предложение).
3. Выявление ключевых навыков, требуемых для вакансий.

**Анализ работодателей:**
1. Определение самых активных работодателей (по количеству вакансий).
2. Анализ предпочитаемых сфер деятельности и регионов для размещения вакансий.
3. Изучение стратегий привлечения сотрудников различными компаниями.

**Анализ регионов:**
1. Идентификация регионов с наибольшим количеством вакансий и работодателей.
2. Сравнение зарплат и спроса на специалистов в разных регионах.
3. Выявление перспективных регионов для поиска работы или бизнеса.

**Анализ сфер деятельности:**
1. Определение наиболее востребованных сфер деятельности на рынке труда.
2. Изучение конкуренции в разных сферах.
3. Анализ тенденций развития различных отраслей.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
# текст запроса
query = f"""
    SELECT a.name AS area, COUNT(v.id) AS cnt
    FROM areas a
    JOIN vacancies v ON a.id = v.area_id
    GROUP BY a.name
    ORDER BY cnt DESC
    LIMIT 5;
"""

In [ ]:
# результат запроса
df_top_5_areas = pd.read_sql_query(query, connection)

# Вывод результата
print(df_top_5_areas)

# Закрываем соединение
connection.close()


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
df = pd.read_sql_query("SELECT * FROM vacancies", connection)

In [ ]:
# результат запроса

# Фильтрация DataFrame
filtered_df = df[(df['salary_from'].notna()) | (df['salary_to'].notna())]

# Подсчет вакансий
vacancy_count = filtered_df.shape[0]

# Вывод результата
print(f"Количество вакансий с заполненными полями зарплаты: {vacancy_count}")

# Закрываем соединение
connection.close()

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
# текст запроса
query = f"""
    SELECT *
    FROM vacancies
"""

In [ ]:
# результат запроса
df_vacancies = pd.read_sql_query(query, connection)

# Вычисление средних значений
avg_salary_from = round(df_vacancies['salary_from'].mean())
avg_salary_to = round(df_vacancies['salary_to'].mean())

# Вывод результата с помощью f-строк
print(f"Средняя нижняя граница зарплаты: {avg_salary_from}")
print(f"Средняя верхняя граница зарплаты: {avg_salary_to}")

# Закрываем соединение
connection.close()

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# текст запроса
query = f"""
    SELECT 
        schedule,
        employment,
        COUNT(*) AS cnt
    FROM vacancies
    GROUP BY schedule, employment
    ORDER BY cnt DESC
"""

In [ ]:
# результат запроса
df_vacancies = pd.read_sql_query(query, connection)

# Вывод результата
print(df_vacancies.iloc[1])

# Закрываем соединение
connection.close()

NameError: name 'connection' is not defined

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# текст запроса
query = f"""
SELECT experience, COUNT(*) AS vacancy_count
FROM vacancies
GROUP BY experience
ORDER BY vacancy_count ASC;
"""

In [ ]:
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

NameError: name 'connection' is not defined

***

In [ ]:
# выводы по детальному анализу вакансий

**География спроса:**

1. **Лидеры рынка:** Москва, Санкт-Петербург, Минск, Новосибирск, Алматы - топ-5 городов по количеству вакансий, что указывает на концентрацию спроса в этих регионах.
2. **Актуальность регионального анализа:** Важно учитывать географию при поиске работы и определении стратегии развития карьеры.

**Заработная плата:**

1. **Средняя зарплата:** Заполненные поля с зарплатой в 24073 вакансиях указывает на то, что информация о зарплате достаточно часто указывается.
2. **Диапазон зарплат:** Средняя нижняя граница зарплатной вилки (71065) и верхняя граница (110537) показывают достаточно широкий диапазон зарплат в сфере дата-сайенса.
3. **Факторы влияния на зарплату:** Для более точного анализа необходимо учитывать опыт работы, специализацию, местоположение, размер компании и другие факторы, которые влияют на зарплату.

**Тип рабочего графика и трудоустройства:**

1. **Популярность удалённой работы:** На втором месте по популярности находится комбинация “Удалённая работа - Полная занятость”, что говорит о растущей тенденции к удалённой работе в сфере дата-сайенса.

**Опыт работы:**

1. **Разнообразный опыт:** Вакансии требуют разный опыт работы, от “Нет опыта” до “Более 6 лет”. Это указывает на то, что в сфере дата-сайенса есть возможности для специалистов с разным уровнем опыта.

**Pекомендации:**

1. **Изучить конкретные вакансии в интересующих регионах:** Изучите наиболее популярные вакансии в Москве, Санкт-Петербурге, Минске, Новосибирске и Алматы, чтобы определить спрос на конкретные навыки и специализации.
2. **Уточнить ожидания по зарплате:** Изучите средние зарплаты в разных регионах и сферах деятельности в сфере дата-сайенса, чтобы сформировать реалистичные ожидания.
3. **Рассмотреть возможность удалённой работы:** Если вас интересует удалённая работа, то следует искать вакансии с соответствующим типом рабочего графика.
4. **Развивать специализацию:** Изучите наиболее востребованные навыки в сфере дата-сайенса в вашем регионе, чтобы стать более конкурентоспособным специалистом.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query = f"""
WITH EmployerVacancyCount AS (
    SELECT e.name, COUNT(*) AS vacancy_count
    FROM employers e
    JOIN vacancies v ON e.id = v.employer_id
    GROUP BY e.name
    ORDER BY vacancy_count DESC
)
SELECT name
FROM EmployerVacancyCount
WHERE vacancy_count = (SELECT vacancy_count FROM EmployerVacancyCount ORDER BY vacancy_count DESC LIMIT 1 OFFSET 0)
OR vacancy_count = (SELECT vacancy_count FROM EmployerVacancyCount ORDER BY vacancy_count DESC LIMIT 1 OFFSET 4);
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query = f"""
WITH EmployerVacancyCount AS (
    SELECT e.name, COUNT(*) AS vacancy_count
    FROM employers e
    JOIN vacancies v ON e.id = v.employer_id
    GROUP BY e.name
    ORDER BY vacancy_count DESC
)
SELECT name
FROM EmployerVacancyCount
WHERE vacancy_count = (SELECT vacancy_count FROM EmployerVacancyCount ORDER BY vacancy_count DESC LIMIT 1 OFFSET 0)
OR vacancy_count = (SELECT vacancy_count FROM EmployerVacancyCount ORDER BY vacancy_count DESC LIMIT 1 OFFSET 4);
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
# текст запроса
query = f"""
SELECT MAX(region_count) AS max_region_count
FROM (
    SELECT e.name, COUNT(DISTINCT v.area_id) AS region_count
    FROM employers e
    JOIN vacancies v ON e.id = v.employer_id
    GROUP BY e.name
) AS employer_region_counts;
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# текст запроса
query = f"""
SELECT COUNT(DISTINCT e.id)
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL;
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query = f"""
SELECT e.name
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(DISTINCT ei.industry_id) = 4
ORDER BY e.name
LIMIT 1 OFFSET 2;
"""

In [ ]:
# результат запроса
# Получаем данные из базы данных в виде DataFrame
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query = f"""
SELECT COUNT(DISTINCT e.id)
FROM EMPLOYERS e
JOIN EMPLOYERS_INDUSTRIES ei ON e.id = ei.employer_id
JOIN INDUSTRIES i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод DataFrame
print(df.iloc[0, 0])

# Закрытие соединения с базой данных
connection.close()

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников

# Скачиваем HTML-код страницы Википедии
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
soup = BeautifulSoup(response.text, 'html.parser')
table = soup.find('table', {'class': 'standard sortable'})

# Извлечение заголовков и данных из таблицы
headers = [th.text.strip() for th in table.find('tr').find_all('th')]
data = []
for row in table.find_all('tr')[1:]:
    row_data = [cell.text.strip() for cell in row.find_all('td')]
    data.append(row_data)
df = pd.DataFrame(data, columns=headers)
cities_list = df['Город'].tolist()

# Создаем соединение с базой данных
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# SQL-запрос для получения информации о вакансиях
query = """
SELECT
    a.name AS region,
    COUNT(v.id) AS vacancy_count
FROM
    vacancies v
JOIN
    employers e ON v.employer_id = e.id
JOIN
    areas a ON v.area_id = a.id
WHERE
    e.name = 'Яндекс' AND a.name IN ({})
GROUP BY
    a.name
UNION ALL
SELECT
    'Total' AS region,
    COUNT(v.id) AS vacancy_count
FROM
    vacancies v
JOIN
    employers e ON v.employer_id = e.id
WHERE
    e.name = 'Яндекс'
ORDER BY
    vacancy_count DESC;
""".format(', '.join(["'{}'".format(city) for city in cities_list]))

# Получаем данные из базы данных в виде DataFrame
df = pd.read_sql_query(query, connection)

# Выводим результат
print(df)

# Закрываем соединение с базой данных
connection.close()

***

In [ ]:
# выводы по анализу работодателей

**Лидеры рынка:**

1. **Яндекс и Газпром нефть:** Эти компании занимают первое и пятое места по количеству вакансий, что говорит об их активной роли на рынке труда в сфере дата-сайенса.

**География:**

1. **Россия - главный рынок:** Большинство работодателей (8419) находятся в России, что подтверждает ее ведущую роль в сфере дата-сайенса.
2. **Широкое географическое присутствие:** 181 регион - это значительное количество для одного работодателя, что говорит о широком географическом охвате вакансий и о возможности найти работу в разных регионах.

**Сфера деятельности:**

1. **Разработка ПО - лидер:** “Разработка программного обеспечения” - самая популярная сфера деятельности среди работодателей (3553 компаний). Это указывает на высокий спрос на дата-сайентистов в IT-сфере.
2. **Отсутствие информации:** 8419 работодателей не указали сферу деятельности. Это может быть связано с ошибками в базе данных, но также может указывать на то, что некоторые компании не делают акцент на сферу деятельности при поиске дата-сайентистов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
# текст запроса
query = f"""
SELECT
    COUNT(Vacancies.id) AS VacanciesCount
FROM Vacancies
WHERE LOWER(Vacancies.name) LIKE '%data%' OR LOWER(Vacancies.name) LIKE '%данн%';
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(df)

# Закрытие подключения
connection.close()

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# текст запроса
query = f"""
SELECT COUNT(*) AS VacanciesCount
FROM Vacancies
WHERE (
  LOWER(Vacancies.name) LIKE '%data scientist%' OR
  LOWER(Vacancies.name) LIKE '%data science%' OR
  LOWER(Vacancies.name) LIKE '%исследователь данных%' OR
  LOWER(Vacancies.name) LIKE '%ml%' AND LOWER(Vacancies.name) NOT LIKE '%html%' OR
  LOWER(Vacancies.name) LIKE '%machine learning%' OR
  LOWER(Vacancies.name) LIKE '%машинн%обучен%'
)
AND (
  LOWER(Vacancies.name) LIKE '%junior%' OR
  Vacancies.experience = 'Нет опыта' OR
  Vacancies.employment = 'Стажировка'
);
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, con=connection)  # Указываем подключение здесь

# Вывод результата
print(df)

# Закрытие подключения
connection.close()

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# текст запроса
query = """
SELECT COUNT(*)
FROM Vacancies
WHERE (
  LOWER(Vacancies.name) LIKE '%data scientist%' OR
  LOWER(Vacancies.name) LIKE '%data science%' OR
  LOWER(Vacancies.name) LIKE '%исследователь данных%' OR
  LOWER(Vacancies.name) LIKE '%ml%' AND LOWER(Vacancies.name) NOT LIKE '%html%' OR
  LOWER(Vacancies.name) LIKE '%machine learning%' OR
  LOWER(Vacancies.name) LIKE '%машинн%обучен%'
)
AND (
  LOWER(Vacancies.key_skills) LIKE '%sql%' OR
  LOWER(Vacancies.key_skills) LIKE '%postgres%'
);
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(df)

# Закрытие подключения
connection.close()

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# текст запроса
query = """
SELECT COUNT(*)
FROM Vacancies
WHERE (
  LOWER(Vacancies.name) LIKE '%data scientist%' OR
  LOWER(Vacancies.name) LIKE '%data science%' OR
  LOWER(Vacancies.name) LIKE '%исследователь данных%' OR
  LOWER(Vacancies.name) LIKE '%ml%' AND LOWER(Vacancies.name) NOT LIKE '%html%' OR
  LOWER(Vacancies.name) LIKE '%machine learning%' OR
  LOWER(Vacancies.name) LIKE '%машинн%обучен%'
)
AND (
  LOWER(Vacancies.key_skills) LIKE '%python%'
);
"""

In [ ]:
# результат запроса
df = pd.read_sql_query(query, connection)

# Вывод результата
print(df)

# Закрытие подключения
connection.close()

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
# текст запроса
import psycopg2
import pandas as pd

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Создание подключения
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# SQL-запрос с использованием f-строк
query = f"""
SELECT ROUND(AVG(LENGTH(Vacancies.key_skills) - LENGTH(REPLACE(Vacancies.key_skills, CHR(9), '')) + 1), 2) AS average_key_skills
FROM Vacancies
WHERE (
  LOWER(Vacancies.name) LIKE '%data scientist%' OR
  LOWER(Vacancies.name) LIKE '%data science%' OR
  LOWER(Vacancies.name) LIKE '%исследователь данных%' OR
  LOWER(Vacancies.name) LIKE '%ml%' AND LOWER(Vacancies.name) NOT LIKE '%html%' OR
  LOWER(Vacancies.name) LIKE '%machine learning%' OR
  LOWER(Vacancies.name) LIKE '%машинн%обучен%'
)
AND NOT (
  LOWER(Vacancies.name) LIKE '%junior%' OR
  LOWER(Vacancies.experience) = 'нет опыта' OR
  LOWER(Vacancies.employment) = 'стажировка'
)
AND Vacancies.key_skills IS NOT NULL
AND NOT (
  Vacancies.key_skills LIKE '%  %' OR 
  Vacancies.key_skills LIKE '%,  %' OR 
  Vacancies.key_skills LIKE '  ,%'
)
AND NOT (Vacancies.key_skills LIKE '%,%' AND Vacancies.key_skills NOT LIKE '%,  %' AND Vacancies.key_skills NOT LIKE '%,  %');
"""


In [ ]:
# результат запроса
# Получение результатов в датафрейм
df = pd.read_sql_query(query, connection)

# Вывод результата
print(df)

# Закрытие подключения
connection.close()

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса
import psycopg2
import pandas as pd

# Параметры подключения
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# Создание подключения
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# SQL-запрос
query = """
SELECT
  v.experience,
  ROUND(AVG(COALESCE(
    (v.salary_from + v.salary_to) / 2,
    v.salary_from,
    v.salary_to
  ))) AS average_salary
FROM vacancies v
WHERE (
    (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
  )
  AND (v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL)
  AND NOT (
    LOWER(v.name) LIKE '%junior%' OR
    LOWER(v.experience) = 'нет опыта' OR
    LOWER(v.employment) = 'стажировка'
  )
GROUP BY v.experience
ORDER BY v.experience;
"""

# Получение результата в датафрейм
df = pd.read_sql_query(query, connection)



In [ ]:
# результат запроса
# Вывод результата
print(df)

# Вывод средней зарплаты для опыта от 3 до 6 лет
average_salary_3_6 = df[df['experience'] == 'От 3 до 6 лет']['average_salary'].iloc[0]
print(f"Средняя зарплата для опыта 'От 3 до 6 лет': {average_salary_3_6}")

# Закрытие подключения
connection.close()

In [ ]:
# выводы по предметному анализу

1. Акцент на SQL и Python: Сосредоточьтесь на изучении SQL, PostgreSQL и Python, поскольку они являются ключевыми навыками для большинства вакансий.
2. Развитие дополнительных навыков: Изучайте другие востребованные навыки для дата-сайентистов, такие как машинное обучение, глубокое обучение, статистический анализ, визуализация данных и т.д.
3. Поиск практики и проектов: Получите практический опыт в сфере дата-сайенса, реализуя проекты или участвуя в конкурсах по данным.
4. Создание портфолио: Соберите свои работы, проекты и решения в портфолио, чтобы продемонстрировать свой опыт и навыки потенциальным работодателям.


# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

**Рынок труда для дата-сайентистов:**

1. **Высокий спрос:** В базе данных 1771 вакансия связана с обработкой данных, что свидетельствует о высоком спросе на специалистов в этой сфере.
2. **Конкуренция:** Существует 51 вакансия для начинающих дата-сайентистов, что означает высокую конкуренцию за эти позиции.
3. **Ключевые навыки:** SQL/PostgreSQL и Python являются ключевыми навыками для большинства вакансий.
4. **Опыт:** Большинство вакансий требуют опыт работы от 3 до 6 лет, что означает, что начинающим дата-сайентистам будет нелегко найти работу без опыта.
5. **География:** Москва, Санкт-Петербург, Минск, Новосибирск и Алматы - топ-5 городов по количеству вакансий.
6. **Заработная плата:** Средняя нижняя граница зарплатной вилки 71065, верхняя - 110537.
7. **Удалённая работа:** “Удалённая работа - Полная занятость” - второй по популярности тип рабочего графика и трудоустройства.

**Выводы:**
1. Рынок труда для дата-сайентистов динамично развивается.
2. Спрос на специалистов высок, но конкуренция также высокая.
3. Ключевые навыки и опыт являются важнейшими факторами для успешного поиска работы в этой сфере.
4. Изучение трендов и прогнозирование будущих требований рынка труда является важным этапом в развитии карьеры в сфере дата-сайенса.